File: clients/ios/PhotoMechaCore/Sources/PhotoMechaCore/ApiError.swift
import Foundation

/// docs/11 エラー時遷移: API エラーをユーザー導線種別へマッピングする。
///
/// | 種別 | 導線 |
/// |---|---|
/// | unauthorized | トークン破棄 → 再登録 (S00) |
/// | duplicate | 再撮影 (S02) |
/// | unsafeCapture | reason 別に再撮影 / 選択やり直し |
/// | clientOutdated | アプリ更新案内（feature_mismatch / unsupported_algo_version） |
/// | quotaExceeded | ホームへ（翌日回復の案内） |
/// | invalid / notFound / server / network | リトライ導線 |
public enum ApiErrorKind: Sendable, Equatable {
    case unauthorized
    case duplicate
    case unsafeCapture
    case clientOutdated
    case quotaExceeded
    case invalid
    case notFound
    case server
    case network
}

public struct ApiError: Error, Sendable {
    public let kind: ApiErrorKind
    public let statusCode: Int
    public let reason: String?
    public let message: String

    public init(kind: ApiErrorKind, statusCode: Int, reason: String?, message: String) {
        self.kind = kind
        self.statusCode = statusCode
        self.reason = reason
        self.message = message
    }
}

public enum ApiErrorMapper {

    private struct ErrorEnvelope: Decodable {
        let detail: Detail?

        enum Detail: Decodable {
            case text(String)
            case object(error: String?, reason: String?)

            init(from decoder: Decoder) throws {
                if let container = try? decoder.singleValueContainer(),
                   let text = try? container.decode(String.self) {
                    self = .text(text)
                    return
                }
                let container = try decoder.container(keyedBy: CodingKeys.self)
                self = .object(
                    error: try container.decodeIfPresent(String.self, forKey: .error),
                    reason: try container.decodeIfPresent(String.self, forKey: .reason)
                )
            }

            enum CodingKeys: String, CodingKey {
                case error, reason
            }
        }
    }

    public static func map(statusCode: Int, body: Data?) -> ApiError {
        var errorCode: String?
        var reason: String?
        var detailText = ""
        if let body {
            detailText = String(data: body, encoding: .utf8) ?? ""
            if let envelope = try? JSONDecoder().decode(ErrorEnvelope.self, from: body) {
                switch envelope.detail {
                case .object(let error, let objectReason):
                    errorCode = error
                    reason = objectReason
                case .text(let text):
                    detailText = text
                case nil:
                    break
                }
            }
        }

        let kind: ApiErrorKind
        switch (statusCode, errorCode) {
        case (401, _): kind = .unauthorized
        case (404, _): kind = .notFound
        case (409, _): kind = .duplicate
        case (429, _): kind = .quotaExceeded
        case (422, "unsafe_capture"): kind = .unsafeCapture
        case (422, "feature_mismatch"), (422, "unsupported_algo_version"): kind = .clientOutdated
        case (400, _), (422, _): kind = .invalid
        default: kind = .server
        }
        return ApiError(
            kind: kind,
            statusCode: statusCode,
            reason: reason ?? errorCode,
            message: "API error \(statusCode): \(errorCode ?? detailText)"
        )
    }
}

File: clients/android/core/src/main/kotlin/com/photomecha/core/api/ApiError.kt
package com.photomecha.core.api

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive

/**
 * docs/11 エラー時遷移: API エラーをユーザー導線種別へマッピングする。
 *
 * | 種別 | 導線 |
 * |---|---|
 * | UNAUTHORIZED | トークン破棄 → 再登録 (S00) |
 * | DUPLICATE | 再撮影 (S02) |
 * | UNSAFE_CAPTURE | reason 別に再撮影 / 選択やり直し |
 * | CLIENT_OUTDATED | アプリ更新案内（feature_mismatch / unsupported_algo_version） |
 * | QUOTA_EXCEEDED | ホームへ（翌日回復の案内） |
 * | INVALID / NOT_FOUND / SERVER / NETWORK | リトライ導線 |
 */
enum class ApiErrorKind {
    UNAUTHORIZED,
    DUPLICATE,
    UNSAFE_CAPTURE,
    CLIENT_OUTDATED,
    QUOTA_EXCEEDED,
    INVALID,
    NOT_FOUND,
    SERVER,
    NETWORK,
}

class ApiException(
    val kind: ApiErrorKind,
    val statusCode: Int,
    val reason: String? = null,
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause)

object ApiErrorMapper {
    private val json = Json { ignoreUnknownKeys = true }

    fun map(statusCode: Int, body: String?): ApiException {
        val detail = runCatching { json.decodeFromString<ErrorEnvelope>(body ?: "").detail }.getOrNull()
        val detailObject = detail as? JsonObject
        val errorCode = (detailObject?.get("error") as? JsonPrimitive)?.content
        val reason = (detailObject?.get("reason") as? JsonPrimitive)?.content
        val detailText = (detail as? JsonPrimitive)?.content ?: body.orEmpty()

        val kind = when {
            statusCode == 401 -> ApiErrorKind.UNAUTHORIZED
            statusCode == 404 -> ApiErrorKind.NOT_FOUND
            statusCode == 409 -> ApiErrorKind.DUPLICATE
            statusCode == 429 -> ApiErrorKind.QUOTA_EXCEEDED
            statusCode == 422 && errorCode == "unsafe_capture" -> ApiErrorKind.UNSAFE_CAPTURE
            statusCode == 422 && (errorCode == "feature_mismatch" || errorCode == "unsupported_algo_version") ->
                ApiErrorKind.CLIENT_OUTDATED
            statusCode == 400 || statusCode == 422 -> ApiErrorKind.INVALID
            statusCode >= 500 -> ApiErrorKind.SERVER
            else -> ApiErrorKind.SERVER
        }
        return ApiException(
            kind = kind,
            statusCode = statusCode,
            reason = reason ?: errorCode,
            message = "API error $statusCode: ${errorCode ?: detailText}",
        )
    }
}
